src/notebooks/controllers/commands/serverConnectionControllerCommands.ts

```diff
@@ -0,0 +1,307 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+'use strict';
+import { inject, injectable } from 'inversify';
+import { IExtensionSingleActivationService } from '../../../platform/activation/types';
+import { ICommandManager, IVSCodeNotebook } from '../../../platform/common/application/types';
+import {
+    Commands,
+    InteractiveWindowView,
+    JupyterNotebookView,
+    JVSC_EXTENSION_ID
+} from '../../../platform/common/constants';
+import { ContextKey } from '../../../platform/common/contextKey';
+import { IDisposableRegistry, IsWebExtension } from '../../../platform/common/types';
+import { JupyterServerSelector } from '../../../kernels/jupyter/serverSelector';
+import { createDeferred } from '../../../platform/common/utils/async';
+import { IControllerLoader, IControllerRegistration, IVSCodeNotebookController } from '../types';
+import {
+    IMultiStepInput,
+    IMultiStepInputFactory,
+    InputFlowAction,
+    InputStep
+} from '../../../platform/common/utils/multiStepInput';
+import { EventEmitter, QuickPickItem, QuickPickItemKind } from 'vscode';
+import { noop } from '../../../platform/common/utils/misc';
+import { isLocalConnection } from '../../../kernels/types';
+import { IJupyterServerUriStorage, IServerConnectionType } from '../../../kernels/jupyter/types';
+import { DataScience } from '../../../platform/common/utils/localize';
+
+function groupBy<T>(data: ReadonlyArray<T>, compare: (a: T, b: T) => number): T[][] {
+    const result: T[][] = [];
+    let currentGroup: T[] | undefined = undefined;
+    for (const element of data.slice(0).sort(compare)) {
+        if (!currentGroup || compare(currentGroup[0], element) !== 0) {
+            currentGroup = [element];
+            result.push(currentGroup);
+        } else {
+            currentGroup.push(element);
+        }
+    }
+    return result;
+}
```

Comment:
Ohh fancy! Would it make sense to have a common place for functional utilities like this?

Reply:
Maybe. Not sure where that would be. It's only used in this class at the moment.